[FrameworkBundle][HttpKernel] Add deprecation warning to show `HttpKernel::handle()` will catch throwables

Nyholm opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 6.2
| Bug fix?      | no
| New feature?  | no
| Deprecations? | yes
| Tickets       | Fix #16205
| License       | MIT
| Doc PR        | symfony/symfony-docs#... <!-- required for new features -->

I suggest that starting from Symfony 7.0 the `HttpKernel` will start to catch `\Throwable` and convert them to a `Response`.

This was first asked in #16205, I face a similar issue with Runtime component and Bref.. 

----------

The reason I push for this change is to embrace the request/response workflow of the Kernel without trusting the custom error handler. In an environment where you serve multiple requests with the same PHP process (read: RoadRunner, Swoole, Bref) you would write something like: 

```php
$kernel = new Kernel('prod', false);
while (true) {
  $request = /* create sf request from custom environment */
  try {
    $response = $kernel->handle(request);
    return ResponseConverter::convert($response);
  } catch (\Throwable $e) {
    exit(1);
  }
}
```
(pseudo code of course. Here is a [real example](https://github.com/php-runtime/bref/blob/0.3.1/src/BrefRunner.php#L30-L43))

The `exit(1)` means a hard crash. For Bref Runtime it would result in a 500 error from API-gateway. Since the `\Throwable` is caught, the Symfony error handler is not used. If we would not to catch the `\Throwable`, then the Symfony error handler would be used, but it would print a Response instead of returning it. (Printing a response will just add HTML on the CLI...)

-----------

Other PRs and issues related to this: 

- https://github.com/symfony/symfony/issues/45301
- https://github.com/symfony/symfony/pull/26514
- https://github.com/symfony/symfony/issues/22128
- https://github.com/symfony/symfony/pull/36885
- https://github.com/symfony/symfony/issues/25467

I'm happy to let the `HttpKernel` to catch the `\Throwable` exception right now, but I thought this very conservative PR would have a higher change to get merged. 

Also note that we do not specify any behaviour on our [HttpKernelInterface](https://github.com/symfony/symfony/blob/v6.0.7/src/Symfony/Component/HttpKernel/HttpKernelInterface.php#L43)


-------

To remove the deprecation message you need to add this to your config: 

```yaml
framework:
    catch_all_throwables: true
```

Nyholm committed:
Add deprecation warning to show HttpKernel::handle() will catch throwables

stof commented:
A deprecation for which there is no way to opt-in for the new behavior is bad: it does not help migrating and will annoy people seeing it without being able to get rid of it.

Nyholm commented:
Hm. You are very correct. Thank you. Do you have a suggestion on how to make an opt-in?

stof commented:
Well, you would need some kind of configuration parameter that can turn on the catching of errors.

Nyholm commented:
Thank you for the review. I've update the PR with a way to opt-in to the new behaviour. I've also added some tests.

Nyholm commented:
Psalm and Fabbot are both false negatives. It seams like the 8.2 test fails for unrelated reasons. 

PHPUnit 8.1 low-deps fails because of this PR. Any suggestions how I could fix them?

derrabus commented:
Thank you very much for bringing this topic up again. I like how you allow to opt-in to the new behavior.

> PHPUnit 8.1 low-deps fails because of this PR. Any suggestions how I could fix them?

The lazy fix would be to add conflict rules to the composer.json files of SecurityBundle and WebProfilerBundle, for FrameworkBundle < 6.1. The alterantive would be to make the test setup for the bundles more complex by setting/removing that option conditionally and I'm not sure that's worth it.

Nyholm commented:
Thank you for the feedback. I've updated the PR.

Nyholm commented:
The PR is rebased and updated. 

status: Waiting for review

Nyholm commented:
The PR is updated for 6.2. Fabbot is a false negative and the PHP 8.2 test fail is unrelated. 

I would be happy to get another review on this. I believe it is ready to merge

derrabus commented on src/Symfony/Bundle/FrameworkBundle/Test/ExceptionListener.php:
```diff
@@ -0,0 +1,41 @@
+ */
+class ExceptionListener
+{
+    private static array $exceptions = [];
```
```suggestion
    /**
     * @var list<\Throwable>
     */
    private static array $exceptions = [];
```

derrabus commented on src/Symfony/Bundle/FrameworkBundle/Test/ExceptionListener.php:
```diff
@@ -0,0 +1,41 @@
+        self::$exceptions[] = $event->getThrowable();
+    }
+
+    public static function getExceptions(): array
```
```suggestion
    /**
     * @return list<\Throwable>
     */
    public static function getExceptions(): array
```

Nyholm commented:
Thank you for the review. I've applied the changes.

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -133,6 +133,7 @@ public function getConfigTreeBuilder(): TreeBuilder
                 ->scalarNode('error_controller')
                     ->defaultValue('error_controller')
                 ->end()
+                ->booleanNode('catch_throwable_exception')->defaultFalse()->info('Tells HttpKernel to catch every kind of Throwable')->end()
```
Maybe we could/should rename `catch_throwable_exception` to something like `catch_all_throwables` as not all throwables are exceptions.

Nyholm replied:
Yeah. I was hesitant of this name too. I like your suggestion

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Test/ExceptionListener.php:
```diff
@@ -0,0 +1,47 @@
+
+/**
+ * This listener allows you to inspect all exceptions thrown by the application.
+ * This is useful since the HttpKernel catches all exceptions and turn them into
```
```suggestion
 * This is useful since the HttpKernel catches all exceptions and turns them into
```

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Test/ExceptionListener.php:
```diff
@@ -0,0 +1,47 @@
+    public static function getExceptions(): array
+    {
+        $exceptions = self::$exceptions;
+        self::$exceptions = [];
```
For a getter method I would not expect the method to modify the object's state but to be able to call the method several times with the same result.

Nyholm replied:
I'll name this `resetExceptions()`

xabbuh replied:
Now it's not really clear that I could use this method retrieve the list of all caught exceptions. What about adapting the schema used for methods in the SPL and use something like `shiftAll()`?

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/Uid/config_disabled.yml:
```diff
@@ -5,3 +5,8 @@ framework:
+services:
+    Symfony\Bundle\FrameworkBundle\Test\ExceptionListener:
+        tags:
+            - { name: kernel.event_listener, event: kernel.exception }
```
Would it make sense instead to let the `ExceptionListener` class implement the `EventSubscriberInterface`? That would make the registration in application easier.

Nyholm replied:
It is just a class for testing. I think it should be opt-in.

xabbuh replied:
It's still opt-in as it won't do anything if I don't register it as a service in my application, right? But assuming a standard Symfony application I would only have this configuration code instead:

```yaml
services:
    Symfony\Bundle\FrameworkBundle\Test\ExceptionListener: ~
```

Nyholm replied:
Thank you. 

I've updated the code

Nyholm replied:
I had to make it: 

```yaml
services:
    Symfony\Bundle\FrameworkBundle\Test\ExceptionSubscriber:
        tags:
            - {name: kernel.event_subscriber}

```

Because EventSubscriberInterface was not automatically added the tag in this test env.

xabbuh commented on src/Symfony/Bundle/FrameworkBundle/Tests/Functional/UidTest.php:
```diff
@@ -31,12 +32,16 @@ protected function setUp(): void
+        $this->assertSame(500, $client->getResponse()->getStatusCode());
+        $exceptions = ExceptionListener::getExceptions();
+        $this->assertCount(1, $exceptions);
+        $exception = $exceptions[array_key_first($exceptions)];
```
```suggestion
        $exception = reset($exceptions);
```

Nyholm commented:
I've applied all changes and the CI is green

HeahDude commented on src/Symfony/Component/HttpKernel/HttpKernel.php:
```diff
@@ -82,6 +87,18 @@ public function handle(Request $request, int $type = HttpKernelInterface::MAIN_R
                 throw $e;
             }
 
+            return $this->handleThrowable($e, $request, $type);
```
Wouldn't it reduce the diff while keeping only one `catch` block to do something like this:
```php
try {
    // ...
} catch (\Throwable $e) {
    // only the first "if" needs to be removed in 7.0
    if (!$e instanceof \Exception && !$this->catchThrowable) {
        throw $e;
    }
    if ($e instanceof RequestExceptionInterface) {
        $e = new BadRequestHttpException($e->getMessage(), $e);
    }
    if (false === $catch) {
        $this->finishRequest($request, $type);

        throw $e;
    }

    return $this->handleThrowable($e, $request, $type);
}
```
?

Nyholm replied:
Not by a hole not, no. I don’t think so. (I’ll try). 

I made two catch blocks to explicitly show this new code path. (And to be 100% sure I (and anybody after me) don’t make mistakes with the exception code path)

HeahDude replied:
I added a comment in my example as I am talking about both the current diff and the one we'll get with 7.0.

Nyholm replied:
Thank you.

HeahDude commented on src/Symfony/Component/HttpKernel/HttpKernel.php:
```diff
@@ -238,6 +255,12 @@ private function handleThrowable(\Throwable $e, Request $request, int $type): Re
             return $this->filterResponse($response, $request, $type);
         } catch (\Exception) {
+            return $response;
+        } catch (\Throwable $e) {
```
Same question here.

Nyholm commented:
I've applied the suggested changes. I've rebased the PR and the tests are green. 

Could I get a final round of reviews?

HeahDude commented on src/Symfony/Bundle/FrameworkBundle/Test/ExceptionSubscriber.php:
```diff
@@ -0,0 +1,56 @@
+        return $exceptions;
+    }
+
+    public static function getSubscribedEvents(): array
```
Missing `{@inheritdoc}`?

Nyholm replied:
Of the 80+ event subscribers, we only add this comment on a handful. I suggest to not include it. It brings little to no value.

HeahDude replied:
Personally I like to use `{@inheritdoc}` since in reviews like this, it's the only way to differentiate methods that implement interfaces or override parents, from the others. (I agree I don't really need help for this scepcific method though, but I mean in general).

HeahDude commented on src/Symfony/Component/HttpKernel/HttpKernel.php:
```diff
@@ -54,13 +54,18 @@ class HttpKernel implements HttpKernelInterface, TerminableInterface
         $this->argumentResolver = $argumentResolver ?? new ArgumentResolver();
+        $this->catchThrowable = $catchThrowable;
+        if (!$catchThrowable) {
+            trigger_deprecation('symfony/http-kernel', '6.2', 'Starting from 7.0, "%s::handle()" will catch \Throwable exceptions and convert them to HttpFoundation responses.', self::class);
```
This depreciation message is not actionable. It should tell to pass `true` as last argument to get rid of it.

Although this would still not be the right fix in the framework context, since the `catch_all_throwables` option needs to be set to `true`.
Maybe the configuration needs a `->validate()` node to trigger on `false` with the "good" message.

Nyholm replied:
I've updated the exception message. 

> Maybe the configuration needs a ->validate() node to trigger on false with the "good" message.

Hm.. Yeah. But isn't that overkill?

HeahDude replied:
I can see two paths. Once you get rid of the deprecation you should be ready for 7.0, but in this case it means passing `true` instead of `false`.

 - Removing the argument and the config node could be done in 8.0, once first deprecated in 7.1.

 - However removing the arg as it seems to be expected in 7.0 won't break, but will make passing the arg some dead code without notice. Also removing the config node in 7.0 would be a BC break if devs continue to define it to `true`.

Nyholm replied:
Since the constructor argument and config is part of the deprecation layer they will both be removed in 7.0. 
Aren't we overthinking now?

HeahDude replied:
`choices_as_values` option was used for Symfony 2 deprecation layer, but deprecated in 3.1,
`logout_on_user_change` option was used in Symfony 3 deprecation layer and deprecated in 4.1 and so on.

The BC should be to not change your code when upgrading to 7.0, in this case changing `false` to `true` would make it work, but once done, another change would be required to remove config and arg, so they must still exist before we can safely remove them.

In other words, the deprecation layer itself cannot be deprecated in 6.x nor removed in 7.0, it needs to be deprecated in 7.1.

HeahDude replied:
>Maybe the configuration needs a ->validate() node to trigger on false with the "good" message.

>Hm... Yeah. But isn't that overkill?

Well, the problem I see is that most devs using the framework do not even know there is an `HttpKernel` service nor how to change this arg.
However, mentioning the config key in this message from the component feels wrong.

HeahDude commented on src/Symfony/Component/HttpKernel/HttpKernel.php:
```diff
@@ -54,13 +54,18 @@ class HttpKernel implements HttpKernelInterface, TerminableInterface
         $this->resolver = $resolver;
         $this->requestStack = $requestStack ?? new RequestStack();
         $this->argumentResolver = $argumentResolver ?? new ArgumentResolver();
+        $this->catchThrowable = $catchThrowable;
```
Does it need to be deprecated as well, or it is allowed to be `false` in 7.0?
The current message sounds like it won't be supported anymore.

Nyholm replied:
hm.. Im not sure we need to deprecate the property as it is part of the deprecation layer already.

HeahDude reviewed (commented):
~It would seem the `composer.json` file of the `WebProfilerBundle` needs an update as well (https://github.com/symfony/symfony/blob/6.2/src/Symfony/Bundle/WebProfilerBundle/composer.json#L21).~

EDIT: it seems I've missed https://github.com/symfony/symfony/pull/45997/files#diff-b7944e9265a9fd4f68f0fb505d9c8375791e607b9709dd3bd93519990683f4c7R62

HeahDude commented:
It would be nice to have this default to `true` with the 6.2 recipe for new projects (and smooth upgrade with `composer recipes:update`).

OskarStark commented on src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php:
```diff
@@ -133,6 +133,7 @@ public function getConfigTreeBuilder(): TreeBuilder
                 ->scalarNode('error_controller')
                     ->defaultValue('error_controller')
                 ->end()
+                ->booleanNode('catch_all_throwables')->defaultFalse()->info('Tells HttpKernel to catch every kind of Throwable')->end()
```
```suggestion
                ->booleanNode('catch_all_throwables')->defaultFalse()->info('HttpKernel will catch all kinds of \Throwable')->end()
```
?

OskarStark commented on src/Symfony/Bundle/FrameworkBundle/Test/ExceptionSubscriber.php:
```diff
@@ -0,0 +1,56 @@
+ * This is useful since the HttpKernel catches all throwables and turns them into
+ * an HTTP response.
+ *
+ * This class should only be used in tests.
```
Should we make it `@internal` or is it not needed in `\Test` sub namespace?

Nyholm replied:
I put it in Test namespace for two reasons: 
1) It is a nice small tool
2) I needed a way to figure out if FrameworkBundle was at version 6.2

OskarStark commented on src/Symfony/Bundle/FrameworkBundle/Tests/Functional/app/Uid/config_disabled.yml:
```diff
@@ -5,3 +5,8 @@ framework:
+services:
+    Symfony\Bundle\FrameworkBundle\Test\ExceptionSubscriber:
+        tags:
+            - {name: kernel.event_subscriber}
```
```suggestion
            - { name: kernel.event_subscriber }
```
As we use whitespaces on the import above 🤷‍♂️

xabbuh commented on src/Symfony/Component/HttpKernel/Tests/HttpKernelTest.php:
```diff
@@ -62,6 +77,80 @@ public function testHandleWhenControllerThrowsAnExceptionAndCatchIsTrueWithAHand
+        $kernel = $this->getHttpKernel($dispatcher, function () { throw new \TypeError('foo'); });
+        $this->expectException(\TypeError::class);
+        $response = $kernel->handle(new Request(), HttpKernelInterface::MAIN_REQUEST, false);
+        $this->assertEquals('foo', $response->getContent());
```
this statement cannot be reached

Nyholm replied:
Thank you

Nyholm commented:
I've rebased the PR and squashed the commits. 

The CI failures are not related to this PR. Could I ask you to review this PR again to decide if it is good enough to merge.

fabpot commented:
Thank you @Nyholm.

Nyholm commented:
Awesome. Thank you for merging. 

Big thank you to all reviewers too
